### ISSUE TITLE
The first request to a MVC route is broken with 1.1 RC

### ISSUE THREAD

**Issue by thegyorf** [Microsoft/aspnet-api-versioning#130]
Navigating to a MVC route on the first request results in the following error:

> System.ArgumentException: Value cannot be null or empty.
> Parameter name: viewPath
>    at Microsoft.AspNetCore.Mvc.ViewEngines.CompositeViewEngine.GetView(String executingFilePath, String viewPath, Boolean isMainPage)
>    at Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor.FindView(ActionContext actionContext, ViewResult viewResult)
>    at Microsoft.AspNetCore.Mvc.ViewResult.<ExecuteResultAsync>d__26.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeResultAsync>d__30.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextResultFilterAsync>d__28.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ResultExecutedContext context)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeNextResourceFilter>d__22.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ResourceExecutedContext context)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
>    at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeAsync>d__20.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.VisualStudio.Web.BrowserLink.BrowserLinkMiddleware.<ExecuteWithFilter>d__7.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7.MoveNext()

Any subsequent request to the route works just fine. API routes, Swashbuckle/Swagger integration, & the SwaggerUI page work just fine only the first request to a MVC route has a problem.

At first I thought it might have something to do with using Swashbuckle/Swagger so I made a fresh ASP.NET Core 1.1 MVC project in Visual Studio 2017, added the 1.1 RC versioning package, updated Startup.cs, and updated the Home controller to use attribute routing but the problem persists.

**Comment by commonsensesoftware** [comment]
Odd. Out of curiosity, how are you requesting the MVC page? Once you add API versioning, everything is versioned, even MVC UI controllers. You might consider applying `[ApiVersionNeutral]` to your UI controllers. Regardless, things shouldn't fail this way. I'll see what I'm able to discover.

**Comment by commonsensesoftware** [comment]
If you're able to share you configuration or even your entire vanilla project, that would be useful in the investigation.

**Comment by thegyorf** [comment]
The controller is ApiVersionNeutral and I'm just browsing to the pages while debugging through Visual Studio.

I've uploaded the project here: https://github.com/thegyorf/IssuesRepro/tree/master/ApiVerIssue/MvcApiTest

**Comment by commonsensesoftware** [comment]
Did things ever work?  Did you ever have the expected behavior?  This might be related to the #133 regression.

**Comment by thegyorf** [comment]
Yes it seems likely that #133 is related.

But no, with the RC build the MVC routes will always fail the first request and then work as expected with every subsequent request.

Reverting to the Beta2 build fixes this issue but brings back this one https://github.com/Microsoft/aspnet-api-versioning/issues/106#issuecomment-299097841 :p

**Comment by commonsensesoftware** [comment]
Ok ... perfect. That's what I was looking for. I may even work some part of your repro into the acceptance tests so this regression does occur again. You'd think with all the tests I have that this type of thing couldn't slip through, but there are a **lot** of application configurations to try and validate.  Thanks for the info.

**Comment by commonsensesoftware** [comment]
[RC2](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Versioning/1.1.0-rc2) has been officially published. My acceptance tests now include a simple MVC view. Let me know if you run into any other issues.

**Comment by thegyorf** [comment]
Updated the project in my repository to RC2 but am still getting the same first request error on the MVC routes.

**Comment by commonsensesoftware** [comment]
Odd. I'll pull down your full repro and investigate.

**Comment by commonsensesoftware** [comment]
I confirmed that this is different, but unresolved bug. There is something illusively different about a route for a view versus an API controller. Regardless, I've tracked down the issue and I'll have a fix for it shortly. In addition to expanding the acceptance tests, I also ran the changes through your repro. All seems well in the universe again. The repro was useful. Thanks.

I'll try to get RC3 out later today.
